facility/list-facility-structure.js

```diff
@@ -0,0 +1,184 @@
+/*
+    This example demonstrates how to list structure of the facility (level - room - asset)
+    
+    It uses 2-legged authentication - this requires that application is added to the account as service.
+*/
+import { createToken } from '../common/auth.js';
+import { TandemClient } from '../common/tandemClient.js';
+import { ColumnFamilies, Encoding, QC } from '../common/utils.js';
+
+// update values below according to your environment
+const APS_CLIENT_ID = 'YOUR_CLIENT_ID';
+const APS_CLIENT_SECRET = 'YOUR_CLIENT_SECRET';
+const FACILITY_URN = 'YOUR_FACILITY_URN';
+
+async function main() {
+    // STEP 1 - obtain token to authenticate subsequent API calls
+    const token = await createToken(APS_CLIENT_ID,
+        APS_CLIENT_SECRET, 'data:read');
+    const client = new TandemClient(() => {
+        return token;
+    });
+
+    // STEP 2 - get facility & facility template
+    const facility = await client.getFacility(FACILITY_URN);
+    // this structure is used to keep structure data
+    const data = {
+        levels: {},
+        rooms: {},
+        assets: {},
+        roomAssetsMap: {}, // map between room and assets
+        roomLevelMap: {} // map between room and level
+    };
+    const modelRooms = [];
+
+    // STEP 3 - collect assets and related room references
+    for (const link of facility.links) {
+        // skip default model
+        if (isDefaultModel(FACILITY_URN, link.modelId)) {
+            continue;
+        }
+        const assets = await client.getTaggedAssets(link.modelId,
+            [ ColumnFamilies.Standard, ColumnFamilies.DtProperties, ColumnFamilies.Refs, ColumnFamilies.Xrefs ]);
+        const modelId = link.modelId;
+
+        for (const asset of assets) {
+            // unique asset key
+            const assetKey = Encoding.toFullKey(asset[QC.Key]);
```

Comment:
You can not be 100% sure all assets are physical here, so I would just use short keys in this case.